### PR TITLE
Improved preference handling / thread safety (DEV)

### PR DESCRIPTION
**Pull request by BMItr**
This PR addresses a potential thread issue accessing/updating the ValueSetStorage. 
It uses HotDataFlow.
StorageTests have been adjusted

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt [hunk 1]
```diff
@@ -19,55 +26,78 @@ import javax.inject.Inject
 class ValueSetsRepository @Inject constructor(
     private val vaccinationServer: VaccinationServer,
     private val valueSetsStorage: ValueSetsStorage,
-    @AppScope private val scope: CoroutineScope
+    @AppScope private val scope: CoroutineScope,
+    dispatcherProvider: DispatcherProvider
 ) {
 
-    val latestValueSet: Flow<VaccinationValueSet> = valueSetsStorage.valueSet.flow.distinctUntilChanged()
+    private fun Flow<VaccinationValueSet>.distinctUntilChangedByHash() = distinctUntilChangedBy { it.hashCode() }
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt:33]: Did you have trouble with just `distinctUntilChanged`? It uses `A == B` which should be okay for data classes.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt [hunk 2]
```diff
@@ -19,55 +26,78 @@ import javax.inject.Inject
...
+        internalData.updateAsync(
+            onUpdate = { updateValueSet(languageCode = languageCode) },
+            onError = { Timber.e(it, "Updating value set failed") }
+        )
+    }
+
+    private suspend fun updateValueSet(languageCode: Locale): VaccinationValueSet {
+        Timber.v("updateValueSet(languageCode=%s)", languageCode)
         var valueSet = vaccinationServer.getVaccinationValueSets(languageCode = languageCode)
+            ?: valueSetsStorage.vaccinationValueSet
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt:69]: Try to stay within the threadsafe execution of `HotDataFlow`, the `valueSetsStorage.vaccinationValueSet` accesses the current data from storage again, the fallback here should be the original data within your execution scope though, i.e.:
`updateAsync.onUpdate { oldData -> newData }`, if there is an error just do ` {oldData -> oldData}`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt [hunk 3]
```diff
@@ -19,55 +26,78 @@ import javax.inject.Inject
...
-            .also { Timber.d("Value set is %s", it) }
-            ?.let { newValueSet -> valueSetsStorage.valueSet.update { newValueSet.toStoredVaccinationValueSet() } }
+        return when (valueSet != null) {
+            true -> valueSet
+            else -> createEmptyValueSet() // Fallback to empty value set
+        }.also { Timber.v("New value set %s", it) }
     }
 
-    fun clear() {
+    private fun createEmptyValueSet(): VaccinationValueSet =
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt:85]: Could be a static value (i.e. companion object)?

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepositoryTest.kt [hunk 4]
```diff
@@ -154,37 +99,15 @@ class ValueSetsRepositoryTest : BaseTest() {
...
+    fun `clear data of server and local storage`() = runBlockingTest2(ignoreActive = true) {
+        createInstance(this).run {
             clear()
 
-            coVerify(exactly = 1) {
+            coVerify() {
                 vaccinationServer.clear()
-                valueSetsStorage.clear()
+                latestValueSet.first() shouldBe emptyValueSetEn
+                valueSetsStorage.vaccinationValueSet = emptyValueSetEn
```
Comment (d4rken) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepositoryTest.kt:109]: Shouldn't `clear()` write `null` to storage to delete the data?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt (hunk 1 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt:33]: DefaultVaccinationValueSet != StoredVaccinationValueSet

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt (hunk 3 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt:85]: all right. will adjust

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepositoryTest.kt (hunk 4 above)
Comment (BMItter) [reply Corona-Warn-App/src/test/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepositoryTest.kt:109]: HotDataFlow doesn't allow nullable values, so I decided to clear the value set by overriding it with an empty value set. I could change the Storage to return an empty value set as default value

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt (hunk 2 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt:69]: "[...]_updateBlocking/update, alle diese update operationen auf dem HotDataFlow Datensatz sind threadsafe_"
Did I get it wrong?

**Review by d4rken — APPROVED** [review]

**Review by chiljamgossow — APPROVED** [review]

**Pull request merged**